grasp (#151)
im-d-team/Dev-Docs: CS/grasp.md

# GRASP - General Responsibility Assignment Software Patterns

직역하자면 일반적으로 책임을 할당할 때 사용할 수 있는 소프트웨어 패턴이다. 말 그대로 책임 할당에 관한 패턴을 이야기한다. 특이한 점은 보통 생각하는 패턴들과는 달리 실체가 없고 추상적인 개념들이다. 따라서 SOLID 원칙과 유사하게 어떤 식으로 책임을 부여해야 할지에 대한 원칙과 철학을 담고 있다.

## 책임이란?

객체지향에서 책임은 매우 중요한 개념이다. RDD(Responsibility Driven Design)에서는 책임을 **특정 작업을 수행하거나, 정보를 알아야 할 의무(an obligation to perform a *task* or know *information*)** 라고 정의한다. 즉 행동(behavior; doing)과 아는것(data; knowing)을 명확히 구분한다. 따라서 객체의 책임은 다음과 같이 정의할 수 있다.

### 행동(Behavior; Doing)

- 스스로 하는 것(doing something itself)

> - 다른 객체와 협력 하는 것(initiate and coordinate actions with other objects)

Reviewer: ```suggestion
- 다른 객체와 협력 하는 것(initiating and coordinating actions with other objects)
```

```java
public class Customer {
	// doing something itself

>     private Customer() {

Reviewer: indentation이 정돈되면 더 좋을 것 같습니다. (이하 예시 코드 동문)

Author: @jigi-kim 수정했습니다.
5ae580f

깃헙 1tab이 8sapce로 설정되어있었네요..

    }
 
    // doing something itself
    public Customer(String email, String name, UniquenessChecker customerUniquenessChecker) {
        this.email = email;
        this.name = name;

>         // initiate and coordinate actions with other objects

Reviewer: ```suggestion
        // initiating and coordinating actions with other objects
```

>         if (customerUniquenessChecker.isUnique(id)) {

Reviewer: `UniquenessChecker`가 실제 존재하는 것인지는 잘 모르지만, 인자값이 이미 존재하면 `isUnique`가 `true` 값을 반환하나요? 함수명만 봤을 때에는 중복되는 것이 없을 때 `true` 값이 반환될 것처럼 생겨서 호기심이 질문 드립니다.

Reviewer: ```suggestion
        if (!customerUniquenessChecker.isUnique(id)) {
```
다른 섹션의 예시에 정답이 있었네요.

Author: 감사합니다 저건 몰랐네요..ㅋㅋㅋ

    }
    
    // doing something itself
    public void addOrder(Order order) {
    }

}
```

### 아는 것(Data; Knowing)

- 객체의 데이터를 감추거나 공개하는 것(private and public object data)
- 연관된 객체의 참조를 연결하는 것(related objects references)
- 해당 객체가 도출해낼 수 있는 것(things it can derive)

```JAVA
public class Customer extends Entity/*knowing*/{    
    @Getter @Setter private Guid id;   // knowing
    @Getter @Setter private String email; // knowing
    @Getter @Setter private String name; // knowing
    @Getter private List<Order> orders; // knowing
```

## GRASP

GRASP은 다음의 9가지 패턴으로 이루어져있다. 각 패턴은 특정 객체가 가져야할 책임이 무엇인지에 대한 것이다. 

1. Information Expert
2. Creator
3. Controller
4. Low Coupling
5. High Cohesion
6. Indirection
7. Polymorphism
8. Pure Fabrication
9. Protected Variations

### Information Expert

객체에 책임을 할당하는 기본 원칙에 관한 것이다. 동작에 필요한 정보가 있는 객체에 책임을 지정해야 한다. 다르게 말하면 동작에 필요한 정보가 있는 객체는 해당 동작을 할 수있는 책임을 지녀야 한다. 반대로 동작에 필요한 정보가 없는 객체는 관련된 동작을 해서는 안된다.

```java
public class Customer {
    private List<Order> orders;
	
	public int getOrdersTotal(Guid orderId) {
		return orders.stream()
            .map(Order::getValue)
            .sum();
	}
}
```

> 위의 예시에서 `Customer` 클래스는모든 고객의 주문(`List<Orrder>)`을 알고 있다. 따라서 `Customer` 클래스가 주문의 총액을 계산하는 것은 당연한 일이다.

Reviewer: ```suggestion
위의 예시에서 `Customer` 클래스는 모든 고객의 주문(`List<Orrder>)`을 알고 있다. 따라서 `Customer` 클래스는 주문의 총액을 계산할 수 있는 책임을 지녀야 한다.
```
위와 같이 표현하는 것은 어떨까요?

### Creator

> 객체를 생성하는 책임에 관한 것이다. 객체 A와 B가 있다고 가정 했을 때, 객체 B가 다음의 조건 중 하나라도 만족한다면 객체 A를 생성하는 책임을 가져야 한다. 만족하는 개수는 많을수록 좋다.

Reviewer 2: ```suggestion
객체를 생성하는 책임에 관한 것이다. 객체 A와 B가 있다고 가정 했을 때, 객체 B가 다음의 조건 중 하나라도 만족한다면 객체 A를 생성하는 책임을 가져야 한다. B가 A를 생성하는 책임을 가진다면 만족하는 개수는 많을수록 좋다.
```

단순히 '만족하는 개수는 많을수록 좋다'라는 표현은 조금 애매한 면이 있는 것 같아 수정 제안드립니다.

Author: @BKJang '조건을'이 목적어로 추가되는게 더 적절해 보이는데 어떠신가요

Reviewer 2: @Dae-Hwa 좋습니다 👍

Reviewer 2: @Dae-Hwa 수정해주실 것으로 믿고 미리 승인하겠습니다~

- B가 A를 포함하거나 복합적으로 집계
- B가 A를 기록
- B가 A를 밀접하게 사용
- B가 A의 초기화 데이터 보유

```java
public class Customer {
    private List<Order> orders; 

    public void addOrder(List<OrderProduct> orderProducts) {	
		Order order = new Order(orderProducts); // Creator

        if (2 <= orders.stream().filter(Order::isOrderedToday).count()) {
            throw new BusinessRuleValidationException("You cannot order more than 2 orders on the same day");
        }

        orders.add(order);

        addDomainEvent(new OrderAddedEvent(order));
    }
}
```

> `Customer` 클래스는 주문을 집계하고 기록하고 사용한다. 무엇보다도 `Customer` 클래스가 주문을 포함한다. 

Reviewer 2: ```suggestion
`Customer` 클래스는 주문을 집계하고 기록하고 사용한다. 무엇보다도 `Customer` 클래스가 주문(`Order`)을 포함한다. 
```

### Controller

> UI 레이어를 넘나들며 시스템 작동의 제어권을 수신하고 조정하는 객체에 관한 것이다. 시스템, 루트객체, 시스템 실행에 필요한 장치, 핵심 서브시스템을 나타내거나 시스템 동작이 발생하는 시나리오(유스케이스)를 표현하는 객체일 경우 컨트롤러의 책임을 가져야 한다.

Reviewer 2: > UI 레이어를 넘나들며 시스템 작동의 제어권을 수신하고 조정하는 객체에 관한 것이다.

단순히 UI 레이어를 넘나드는 경우에만 Controller라고 보면 되는건가요?
대표적으로 웹 혹은 앱 개발에서는 Controller의 역할이 말씀하신게 맞을 것 같긴한데 다른 개발 분야에서는 아닐 수도 있을 것 같아서요.

Author: @BKJang 모바일 앱 이외의 전통적인 응용소프트웨어에서도 사용자 대상의 소프트웨어라면 UI가 필요합니다. 그러한 맥락에서 사용된 것이고, UI가 필요없는 분야에서는 Facade패턴으로 주요 인터페이스들을 래핑하는 것으로 알고 있는데 공부가 부족하여 자세한 설명은 힘들 것 같습니다.
혹시 궁금하시면 찾아보고 발표해주시는것도 좋을 것 같네요 😄😄😄😄😄

```java
public class CustomerOrdersController {
	private OrderService customerOrderService;

	public CustomerOrdersController(OrderService customerOrderService) {
		this.customerOrderService = customerOrderService;
	}

	@PostMapping("/{customerId}/orders")
	public String addCustomerOrder(Guid customerId, HttpServletRequest request)	{
	   ...
	}
}
```

위는 전형적인 Spring MVC의 컨트롤러인데, 좀 더 엄밀히 말하면 [DispatcherServlet](https://docs.spring.io/spring/docs/3.0.0.M4/spring-framework-reference/html/ch15s02.html)이 Front Controller 역할을 해주는 일종의 GRASP 컨트롤러라고 말할 수 있다. 

### Low Coupling

> 결합도(coupling)가 낮게 유지되도록 책임을 할당 해야 한다. 낮은 결합도는 변화의 영향을 줄이고 낮은 의존성과 재사용성을 증가시키는 방법이다. 위의 Controller 예시는 의존성 주입을 이용해 결합도가 낮다.  `OrderService` 인터페이스의 명세가 바뀌지 않으면 이를 구현하는 서비스 객체가 바뀌어도 큰 영향을 받지 않을 것이다.

Reviewer 2: ```suggestion
결합도(coupling)가 낮게 유지되도록 책임을 할당 해야 한다. 낮은 결합도는 변화의 영향을 줄이고 낮은 의존성과 재사용성을 증가시키는 방법이다. 위의 Controller 예시는 의존성 주입을 이용해 결합도가 낮다. `OrderService` 인터페이스의 명세가 바뀌지 않으면 이를 구현하는 서비스 객체가 바뀌어도 큰 영향을 받지 않을 것이다.
```

    private CacheStore cacheStore;

    public ForeignExchange(CacheStore cacheStore) {
        this.cacheStore = cacheStore;
    }

    @Override
    public List<ConversionRate> getConversionRates() {
        ...
    }

    private static List<ConversionRate> getConversionRatesFromExternalApi() {
        // Communication with external API
    }
}
```

### Protected Variations

변화가 많거나 변경의 가능성이 높은 불안정한 객체나 서브시스템은 안정적인 인터페이스로 감쌀 수 있다.

```java
public class CustomerOrdersRepository {

    private JdbcTemplate jdbcTemplate;

    public void setDataSource(DataSource dataSource) {
        this.jdbcTemplate = new JdbcTemplate(dataSource);
    }

    // JDBC-backed implementations of the methods...
}

@Configuration
public class SpringJdbcConfig {
    @Bean
    public DataSource mysqlDataSource() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName("com.mysql.jdbc.Driver");
        dataSource.setUrl("jdbc:mysql://localhost:3306/springjdbc");
        dataSource.setUsername("guest_user");
        dataSource.setPassword("guest_password");
 
        return dataSource;
    }
}
```

> 위의 예시는 `DataSource`라는 안정적인 인터페이스 안에서 데이터 베이스 setup에 대한 책임을 담당한다. 데이터 베이스가 교체되더라도 `DataSource` 객체의 명세를 그대로 사용하기 때문에 이를 구현한 클래스 를 교체하거나 내부 구현의 내용만 바꿔주면 다른 시스템에 피해 없이 작업을 수행할 수 있다.

Reviewer 2: ```suggestion
위의 예시는 `DataSource`라는 안정적인 인터페이스 안에서 데이터 베이스 setup에 대한 책임을 담당한다. 데이터 베이스가 교체되더라도 `DataSource` 객체의 명세를 그대로 사용하기 때문에 이를 구현한 클래스를 교체하거나 내부 구현의 내용만 바꿔주면 다른 시스템에 피해 없이 작업을 수행할 수 있다.
```

# GRASP - General Responsibility Assignment Software Patterns

> 직역하자면 일반적으로 책임을 할당할 때 사용할 수 있는 소프트웨어 패턴이다. 말 그대로 책임 할당에 관한 패턴을 이야기한다. 특이한 점은 보통 생각하는 패턴들과는 달리 실체가 없고 추상적인 개념들이다. 따라서 SOLID 원칙과 유사하게 어떤 식으로 책임을 부여해야 할지에 대한 원칙과 철학을 담고 있다.

Reviewer 3: SOLID 원칙 에 대해 링크를 걸어주면 좀 더 좋을 것 같습니다!

Author: 수정했습니다.

- B가 A를 포함하거나 복합적으로 집계
- B가 A를 기록
- B가 A를 밀접하게 사용
- B가 A의 초기화 데이터 보유

```java
public class Customer {
    private List<Order> orders; 

    public void addOrder(List<OrderProduct> orderProducts) {	
	Order order = new Order(orderProducts); // Creator

        if (2 <= orders.stream().filter(Order::isOrderedToday).count()) {
            throw new BusinessRuleValidationException("You cannot order more than 2 orders on the same day");
        }

        orders.add(order);

        addDomainEvent(new OrderAddedEvent(order));
    }
}
```

`Customer` 클래스는 주문을 집계하고 기록하고 사용한다. 무엇보다도 `Customer` 클래스가 주문(`Order`)을 포함한다. 

### Controller

UI 레이어를 넘나들며 시스템 작동의 제어권을 수신하고 조정하는 객체에 관한 것이다. 시스템, 루트객체, 시스템 실행에 필요한 장치, 핵심 서브시스템을 나타내거나 시스템 동작이 발생하는 시나리오(유스케이스)를 표현하는 객체일 경우 컨트롤러의 책임을 가져야 한다.

```java
public class CustomerOrdersController {
    private OrderService customerOrderService;

    public CustomerOrdersController(OrderService customerOrderService) {
        this.customerOrderService = customerOrderService;
    }

    @PostMapping("/{customerId}/orders")
    public String addCustomerOrder(Guid customerId, HttpServletRequest request)	{
        //...
    }
}
```

> 위는 전형적인 Spring MVC의 컨트롤러인데, 좀 더 엄밀히 말하면 [DispatcherServlet](https://docs.spring.io/spring/docs/3.0.0.M4/spring-framework-reference/html/ch15s02.html)이 Front Controller 역할을 해주는 일종의 GRASP 컨트롤러라고 말할 수 있다. 

Reviewer 3: 여기서 HttpServletRequest가 Front Controller의 역할을 해주는 것인가요?

Author: 아뇨 설명에 나와있는것처럼 `DispatcherServlet` 이 Front Controller 역할을 해주는 것입니다.
![image](https://user-images.githubusercontent.com/24666330/90588245-20764080-e216-11ea-8fd7-b77a026f0af4.png)

본문 내용과는 조금 떨어져있지만 설명해드리자면, 말씀하신 객체는 서블릿 객체의 매개변수로 사용됩니다.
아래 블로그에 정리가 잘 되있는 것 같네요 혹시나 더 좋은 자료 찾으면 다음에 말씀드리겠습니다.
https://tinkerbellbass.tistory.com/1
https://namocom.tistory.com/824

### Low Coupling

결합도(coupling)가 낮게 유지되도록 책임을 할당 해야 한다. 낮은 결합도는 변화의 영향을 줄이고 낮은 의존성과 재사용성을 증가시키는 방법이다. 위의 Controller 예시는 의존성 주입을 이용해 결합도가 낮다. `OrderService` 인터페이스의 명세가 바뀌지 않으면 이를 구현하는 서비스 객체가 바뀌어도 큰 영향을 받지 않을 것이다.

### High Cohension

좋은 객체는 높은 응집력(cohension)을 가져야 한다. 응집력은 객체 내부의 책임이 얼마나 연관성이 있는지에 관한 것이다. 만약 `Customer` 클래스가 주문 뿐만 아니라 상품의 가격을 관리하는 책임을 가지게 된다면 해당 클래스의 응집력은 크게 떨어진다.

### Indirection

> 컴포넌트나 서비스가 직접 연결되지 않도록 중재 역할을 해주는 책임을 갖는 객체를 이용하는 것이다. MVC패턴에서 모델과 뷰 사이에 컨트롤러를 구성하는 것을 예시로 들 수 있다. 혹은 [Mediator Pattern](https://springframework.guru/gang-of-four-design-patterns/mediator-pattern/)을 이용할 수 있다. 객체가 직접 연결되어 있지 않다는 것은 시스템을 유연하게 만드는 것을 도와줄 수 있지만, 가독성과 추론을 어렵게 만들 수 있는 trade off가 있기 때문에 직접 구현하게 된다면 이를 잘 고려해야 한다.

Reviewer 3: 시스템을 유연하게 만들 수 있다는 것이 무슨 의미인지 궁금합니다

Author: MVC패턴에서 모델과 뷰가 컨트롤러를 거쳐서 연결되는 것처럼 생각하시면 될 것 같습니다.
모델과 뷰는 서로 독립적이죠. 그렇기 때문에 서로 간섭없이 유연한 개발에 도움이 됩니다.

### Polymorphism

유형에 따라 서로 다른 케이스를 처리하고 싶을 경우 다형성을 이용할 수 있다. 

```java
private Guid id;

public Customer(string email, string name, UniquenessChecker customerUniquenessChecker) {
    this.Email = email;
    this.Name = name;
 
    if (!customerUniquenessChecker.isUnique(id)) {
        throw new BusinessRuleValidationException("Customer with this email already exists.");
    }
 
    this.AddDomainEvent(new CustomerRegisteredEvent(this));
}
```

`UniquenessChecker`가 `Customer`에 사용될 경우 `CustomerUniquenessChecker` 다른 곳에 사용될 경우 `OtherUniquenessChecker`를 사용하는 식으로 책임은 같지만 구현에 따라 동작이 달라질 경우 유용하게 쓰일 수 있다.

### Pure Fabrication

> 다른 원칙들을 적용하여 높은 응집력과 낮은 결합력을 유지하기 힘들 경우(정확한 책임을 할당하기 힘든 경우) 도메인에 독립적인 클래스와 인터페이스를 만드는 것이 좋을 수 있다.

Reviewer 3: 도메인에 독립적인 클래스와 인터페이스라는 것이 아예 다른 범주의 것을 뜻하는 건가요? 그리고 그렇게 구현하는게 좋은 이유가 괜한 결합도를 증가시킬 수 있기 때문인지 궁금합니다

Author: 이해하기 어렵다면 일단은 유틸성 클래스라고 생각해봐도 좋을 것 같습니다.
위의 예시에서 `ForeignExchange` 객체는 환전 기능을 담당하고 있습니다.
`Customer` 혹은 `Order` 객체와 독립적입니다. 만약 `Customer` 객체 혹은 `Order` 객체에서 환전이 필요하다고해서 해당 객체들의 책임으로 분류하기는 애매한 상황입니다.
이럴 경우 환전 기능을 담당하는 독립적인 클래스를 분리하여 작성하는 것입니다.
예시로 표현이 한계가 있어 아쉽네요. 혹시 이해가 안가신다면 추가적으로 질문 부탁드립니다.

```java
public interface ForeignExchange {
    List<ConversionRate> getConversionRates();
}

public class ForeignExchangeImpl implements ForeignExchange {
    private CacheStore cacheStore;

    public ForeignExchange(CacheStore cacheStore) {
        this.cacheStore = cacheStore;
    }

    @Override
    public List<ConversionRate> getConversionRates() {
        //...
    }

    private static List<ConversionRate> getConversionRatesFromExternalApi() {
        // Communication with external API
    }
}
```

> 

Reviewer 4: 위에 코드가 Pure Fabrication의 예시인가요? 짤막한 설명도 같이 해주시면 이해가 훨씬 쉬울것 같습니다~

Author: @JegalEun 반영해서 수정했습니다.

### Protected Variations

변화가 많거나 변경의 가능성이 높은 불안정한 객체나 서브시스템은 안정적인 인터페이스로 감쌀 수 있다.

```java
public class CustomerOrdersRepository {

    private JdbcTemplate jdbcTemplate;

    public void setDataSource(DataSource dataSource) {
        this.jdbcTemplate = new JdbcTemplate(dataSource);
    }

    // JDBC-backed implementations of the methods...
}

@Configuration
public class SpringJdbcConfig {
    @Bean
    public DataSource mysqlDataSource() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName("com.mysql.jdbc.Driver");
        dataSource.setUrl("jdbc:mysql://localhost:3306/springjdbc");
        dataSource.setUsername("guest_user");
        dataSource.setPassword("guest_password");
 
        return dataSource;
    }
}
```

> 위의 예시는 `DataSource`라는 안정적인 인터페이스 안에서 데이터 베이스 setup에 대한 책임을 담당한다. 데이터 베이스가 교체되더라도 `DataSource` 객체의 명세를 그대로 사용하기 때문에 이를 구현한 클래스를 교체하거나 내부 구현의 내용만 바꿔주면 다른 시스템에 피해 없이 작업을 수행할 수 있다.

Reviewer 3: 이 부분이 잘 이해 되지 않습니다 Protected Variations가 변경 가능성이 높은 객체 / 서브시스템을 안정적인 인터페이스 (여기서는 DataSource) 로 감쌀 수 있다는 것으로 해석되는데 DB가 다른 종류로 바뀌더라도 DataSource 객체의 명세를 그대로 사용한다는 것이 무슨 뜻인지 궁금합니다

Author: `DataSource` 객체를 사용해보셨다면, 해당 객체에 드라이버명, 데이터베이스 주소, 사용자 이름, 비밀번호를 넣는 것은 공통적이지만, 드라이버의 종류에 따라 각각 다른 동작을 하게 된다는 것을 알고 계실텐데요.
예를들어 mysql드라이버를 넣는다면 mysql과 연결하는 기능을 제공할 것이고, oracle드라이버를 연결하면 oracle과 연결하도록 할것입니다.
주소, 이름, 비밀번호도 마찬가지로 변경이 필요할때 다른 객체를 건드릴 필요 없이 `DataSource`에 들어갈 값들만 변경해주면 전체 시스템에 아무런 영향 없이 해당 내용이 변경 가능할것입니다.
